File: app/src/main/java/com/example/myapplication/ui/feed/FeedFragment.java
package com.example.myapplication.ui.feed;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.core.widget.ContentLoadingProgressBar;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.FragmentNavigator;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.example.myapplication.R;
import com.example.myapplication.data.Project;
import com.example.myapplication.di.AppComponentProvider;
import com.xwray.groupie.GroupAdapter;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

public class FeedFragment extends Fragment implements FeedPresenter.View {
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.recyclerView)
    RecyclerView projectRecyclerView;
    @BindView(R.id.progressView)
    ContentLoadingProgressBar progressView;
    @BindView(R.id.errorView)
    TextView errorView;

    private GroupAdapter groupAdapter = new GroupAdapter();

    private BehaviorSubject<Project> clickSubject = BehaviorSubject.create();

    private Disposable disposable;

    @SuppressWarnings("WeakerAccess")
    @Inject
    FeedPresenter presenter;

    FragmentNavigator.Extras extras;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        AppComponentProvider component = (AppComponentProvider) (getActivity().getApplication());
        component.getComponent().build().feedFragment(this);
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_list, container, false);

        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.d("GotoProject", "bind");

        setupRecyclerView();
        disposable = presenter.bind(this);

    }

    @Override
    public void onDetach() {
        disposable.dispose();
        super.onDetach();
    }

    @Override
    public void showProgress() {
        errorView.setVisibility(View.INVISIBLE);
        projectRecyclerView.setVisibility(View.INVISIBLE);
        progressView.show();
    }

    @Override
    public Completable showProjects(List<Project> projectList) {
        return Completable.create(emitter ->
        {
            ArrayList<ProjectItem> itemList = new ArrayList<>();
            progressView.hide();
            for (Project project : projectList) {
                itemList.add(new ProjectItem(project));
            }
            groupAdapter.update(itemList);
            projectRecyclerView.setVisibility(View.VISIBLE);
            emitter.onComplete();
        });
    }

    @Override
    public Observable<Project> projectClick() {
        return clickSubject;
    }

    @Override
    public Completable showEmpty() {
        return Completable.create(emitter ->
        {
            progressView.hide();
            projectRecyclerView.setVisibility(View.INVISIBLE);
            errorView.setVisibility(View.VISIBLE);
            emitter.onComplete();
        });
    }

    @Override
    public void goToProject(Project project) {
        Log.d("GotoProject", "called");
        Bundle bundle = new Bundle();
        bundle.putParcelable("project", project);
        NavHostFragment.findNavController(this).navigate(R.id.project_click, bundle, null);

    }

    private void setupRecyclerView() {
        projectRecyclerView.setLayoutManager(new LinearLayoutManager(this.getContext()));
        projectRecyclerView.setAdapter(groupAdapter);
        projectRecyclerView.addItemDecoration(new DividerItemDecoration(this.getContext(), DividerItemDecoration.VERTICAL));
        groupAdapter.setOnItemClickListener((item, view) -> {
            ProjectItem projectItem = (ProjectItem) item;
            clickSubject.onNext(projectItem.getProject());
        });
    }
}

File: app/src/main/java/com/example/myapplication/ui/feed/FeedFragment.kt
package com.example.myapplication.ui.feed

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.core.view.ViewCompat
import androidx.core.view.doOnPreDraw
import androidx.navigation.findNavController
import androidx.navigation.fragment.FragmentNavigator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.myapplication.R
import com.example.myapplication.di.appComponent
import com.hannesdorfmann.mosby3.mvi.MviFragment
import com.xwray.groupie.GroupAdapter
import com.xwray.groupie.kotlinandroidextensions.ViewHolder
import io.reactivex.subjects.BehaviorSubject
import kotlinx.android.synthetic.main.fragment_list.*
import kotlinx.android.synthetic.main.item_project.view.*
import javax.inject.Inject

class FeedFragment : MviFragment<FeedView, FeedPresenter>(), FeedView {

	private val groupAdapter: GroupAdapter<ViewHolder> = GroupAdapter()
	private var loadSubject: BehaviorSubject<Unit> = BehaviorSubject.create<Unit>()

	@Inject
	lateinit var presenter: FeedPresenter

	override fun createPresenter(): FeedPresenter = presenter

	override fun onCreate(savedInstanceState: Bundle?) {
		appComponent.fragment(this)
		super.onCreate(savedInstanceState)
	}

	override fun onCreateView(
		inflater: LayoutInflater, container: ViewGroup?,
		savedInstanceState: Bundle?
	): View? {
		setHasOptionsMenu(true)
		return inflater.inflate(R.layout.fragment_list, container, false)
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)
		setupRecyclerView()
		postponeEnterTransition()
		view.doOnPreDraw { startPostponedEnterTransition() }

		loadSubject.onNext(Unit)

	}

	override fun loadProjects(): BehaviorSubject<Unit> = loadSubject

	override fun render(vs: FeedViewState) {
		when (vs) {
			is FeedViewState.Loading -> {
				progressView.show()
				recyclerView.visibility = View.INVISIBLE
			}
			is FeedViewState.ProjectList -> {
				progressView.hide()
				recyclerView.visibility = View.VISIBLE
				val itemList = vs.projectList
					.map { ProjectItem(it) }
					.toList()

				groupAdapter.update(itemList)
			}
			is FeedViewState.Error -> {
				progressView.hide()
				recyclerView.visibility = View.INVISIBLE
				errorView.text = vs.msg
				errorView.visibility = View.VISIBLE
			}
		}
	}

	private fun setupRecyclerView() {
		recyclerView.apply {
			layoutManager = LinearLayoutManager(this.context)
			adapter = groupAdapter
			addItemDecoration(DividerItemDecoration(recyclerView.context, DividerItemDecoration.VERTICAL))
		}

		//TODO ideally this would be Passed as an Intent to the Presenter
		groupAdapter.setOnItemClickListener { item, view ->
			// Pass item to Bundle
			val bundle = bundleOf("project" to (item as ProjectItem).project)

			// Add shared element transitions
			val extras = FragmentNavigator.Extras.Builder()
				.addSharedElement(view.logo, ViewCompat.getTransitionName(view.logo)!!)
				.addSharedElement(view.title, "title")
				.build()

			// Push action to navigation controller
			view.findNavController().navigate(R.id.project_click, bundle, null, extras)
		}
	}
}